cli/worker.rs

```diff
@@ -348,55 +360,126 @@ impl CliMainWorker {
   pub async fn run_tests(
     &mut self,
     shuffle: &Option<u64>,
+    filter: TestFilter,
+    fail_fast_tracker: FailFastTracker,
   ) -> Result<(), AnyError> {
-    let promise = {
-      let scope = &mut self.worker.js_runtime.handle_scope();
-      let cb = self.js_run_tests_callback.as_ref().unwrap().open(scope);
-      let this = v8::undefined(scope).into();
-      let options =
-        serde_v8::to_v8(scope, json!({ "shuffle": shuffle })).unwrap();
-      let promise = cb.call(scope, this, &[options]).unwrap();
-      v8::Global::new(scope, promise)
+    let (tests, origin, mut sender) = {
+      let state_rc = self.worker.js_runtime.op_state();
+      let mut state = state_rc.borrow_mut();
+      let tests = std::mem::take(
+        &mut state.borrow_mut::<ops::testing::TestContainer>().0,
+      );
+      let origin = state.borrow::<ModuleSpecifier>().clone();
+      let sender = state.borrow::<TestEventSender>().clone();
+      (tests, origin, sender)
     };
-    self.worker.js_runtime.resolve_value(promise).await?;
+    let unfiltered = tests.len();
+    let (only, no_only): (Vec<_>, Vec<_>) =
+      tests.into_iter().partition(|(d, _)| d.only);
+    let used_only = !only.is_empty();
+    let tests = if used_only { only } else { no_only };
+    let mut tests = tests
+      .into_iter()
+      .filter(|(d, _)| filter.includes(&d.name))
+      .collect::<Vec<_>>();
+    if let Some(seed) = shuffle {
+      tests.shuffle(&mut StdRng::seed_from_u64(*seed));
+    }
+    sender.send(TestEvent::Plan(TestPlan {
+      origin: origin.to_string(),
+      total: tests.len(),
+      filtered_out: unfiltered - tests.len(),
+      used_only,
+    }))?;
+    for (desc, function) in tests {
+      if fail_fast_tracker.should_stop() {
+        break;
+      }
+      if desc.ignore {
+        sender.send(TestEvent::Result(desc.id, TestResult::Ignored, 0))?;
+        continue;
+      }
+      sender.send(TestEvent::Wait(desc.id))?;
+      let earlier = SystemTime::now();
+      let promise = {
+        let scope = &mut self.worker.js_runtime.handle_scope();
+        let cb = function.open(scope);
+        let this = v8::undefined(scope).into();
+        let promise = cb.call(scope, this, &[]).unwrap();
+        v8::Global::new(scope, promise)
+      };
+      let result = self.worker.js_runtime.resolve_value(promise).await?;
```

Comment:
I like this. This gives us a clear path to implementing test timeout, that will be able to handle blocking sync code. We can now spawn a separate "watchdog" thread that will call `IsolateHandle::terminate_execution` if the test doesn't complete before the deadline 👍